cmd/kubeadm/app/cmd/join.go

```diff
@@ -59,38 +59,40 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 
 	var skipPreFlight bool
 	var cfgPath string
+	var discURL string
+	var discFile string
+	var discToken string
 
 	cmd := &cobra.Command{
 		Use:   "join <master address>",
 		Short: "Run this on any machine you wish to join an existing cluster",
 		Run: func(cmd *cobra.Command, args []string) {
-			j, err := NewJoin(cfgPath, args, &cfg, skipPreFlight)
+			j, err := NewJoin(cfgPath, discURL, discFile, discToken, args, &cfg, skipPreFlight)
 			kubeadmutil.CheckErr(err)
 			kubeadmutil.CheckErr(j.Validate())
 			kubeadmutil.CheckErr(j.Run(out))
 		},
 	}
 
 	cmd.PersistentFlags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file")
+	cmd.PersistentFlags().StringVar(&discURL, "discovery-url", discURL, "Discovery URL")
+	cmd.PersistentFlags().StringVar(&discFile, "discovery-file", discFile, "Discovery File path")
+	cmd.PersistentFlags().StringVar(&discToken, "discovery-token", discToken, "Discovery Token")
+	cmd.PersistentFlags().StringVar(&discToken, "token", discToken, "This sets the token for both discovery and bootstrap auth")
 
 	cmd.PersistentFlags().BoolVar(
 		&skipPreFlight, "skip-preflight-checks", false,
 		"skip preflight checks normally run before modifying the system",
 	)
 
-	cmd.PersistentFlags().Var(
```

Comment:
I suggest we remove the discovery struct later in a different PR because it will change a lot of code. I am willing to do that as a follow up to this.